toolchain/lex/lex.cpp

```diff
@@ -0,0 +1,1309 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lex/lex.h"
+
+#include <array>
+
+#include "common/check.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/Support/Compiler.h"
+#include "toolchain/base/value_store.h"
+#include "toolchain/lex/character_set.h"
+#include "toolchain/lex/helpers.h"
+#include "toolchain/lex/numeric_literal.h"
+#include "toolchain/lex/string_literal.h"
+#include "toolchain/lex/tokenized_buffer.h"
+
+#if __ARM_NEON
+#include <arm_neon.h>
+#define CARBON_USE_SIMD 1
+#elif __x86_64__
+#include <x86intrin.h>
+#define CARBON_USE_SIMD 1
+#else
+#define CARBON_USE_SIMD 0
+#endif
+
+namespace Carbon::Lex {
+
+// Implementation of the lexer logic itself.
+//
+// The design is that lexing can loop over the source buffer, consuming it into
+// tokens by calling into this API. This class handles the state and breaks down
+// the different lexing steps that may be used. It directly updates the provided
+// tokenized buffer with the lexed tokens.
+//
+// Because of the internal linkage attribute, this must be in the cpp file where
+// it's used.
+// TODO: Why is this needed?
```

Comment:
Answering your question from the top-level discussion thread and in the comment here --

It was using internal linkage attribute because it was a nested class, and so couldn't be in an anonymous namespace.

Now that it's not a nested class, you can definitely use an anonymous namespace and I agree that's much cleaner.

How much motivation do you think we need here? My hope is that keeping a class in a `.cpp` file and in an anonymous namespace when its only used in that `.cpp` file is our default pattern?

Reply:
Our general project style has been to use `.h` files for large classes, and we'd typically split things out. So I'm really looking to you for an explanation of why this is required... although you're correct it's not really reusable, we do typically put separate declarations of classes into headers, and this file is currently 1.3k LOC which is very large. Under normal conditions, I believe we would be trying to split it. As a consequence, an explanation is somethign I think would be very helpful.

Reply:
And I'm realizing, maybe you still want internal_linkage. The class was written expecting to be a `friend` of `TokenizedBuffer` (which, I'm not going to refactor that, although I'll note things like `SemIR` don't do that).

Reply:
Suggested a comment you can add.